### PR TITLE
new yral feed

### PR DESCRIPTION
**Pull request by abhishek-tripathi-yral**
- new yral feed in canister with new endpoints

### CODE REVIEW

File: src/canister/individual_user_template/src/api/post/send_update_post_cache.rs [hunk 1]
```diff
@@ -37,11 +37,18 @@ pub fn send_update_post_cache(post_id: &u64) {
...
             "update_post_home_feed",
+            (home_feed_index_score_clone.unwrap(),),
+        );
+        let _ = call::notify(
+            post_cache_canister_principal_id,
+            "update_post_yral_feed",
             (home_feed_index_score_item.unwrap(),),
         );
+
     }
```
Comment (komal-rs) [inline src/canister/individual_user_template/src/api/post/send_update_post_cache.rs:52]: This change should also be done in update_scores_and_share_with_post_cache_if_difference_beyond_threshold

File: src/canister/individual_user_template/src/api/post/send_update_post_cache.rs [hunk 2]
```diff
@@ -37,11 +37,18 @@ pub fn send_update_post_cache(post_id: &u64) {
     });
 
     if home_feed_index_score_item.is_some() {
+        let home_feed_index_score_clone = home_feed_index_score_item.clone();
         let _ = call::notify(
             post_cache_canister_principal_id,
             "update_post_home_feed",
+            (home_feed_index_score_clone.unwrap(),),
+        );
+        let _ = call::notify(
```
Comment (komal-rs) [inline src/canister/individual_user_template/src/api/post/send_update_post_cache.rs:46]: should this change be in home_feed_index_score_item or hot_or_not_index_score_item. Once check how the scores are being calculated and which score do we want to keep. I think it should be HotOrNot score but once check and let's discuss.

Comment (komal-rs) [reply src/canister/individual_user_template/src/api/post/send_update_post_cache.rs:46]: might have to implement recalculate_yral_feed_score, but have to check

**Review by komal-sai-yral — COMMENTED** [review]
Few things :
1. Can the post bettings go on forever ? I think it is limited for only a fixed num of slots right now. In that case we need to show the status appropriately in FE. Might have to discuss this with product, do we even want to show such posts. Check trigger_update_hot_or_not_index for some context - we are removing posts older than 48 hrs every hour so that such posts are not in the hotornot index currently
2. A e2e integration test would be really helpful. Where we check proper functionality of the whole flow - create a bunch of posts, check if yral_feed gets them, bet on the posts, increment timer, check if scores make sense, etc.

**Review by komal-sai-yral — COMMENTED** [review]
Do we have to uncomment these tests ?

**Review by komal-sai-yral — COMMENTED** [review]
Reviewed partially. Will complete tom. 

-----------------------

Would be great if we can have one integration test to check the upgradation from SlotId to NewSlotType to avoid canister crashes. 

Steps:
1. save the individual_canister wasm of main branch as old_wasm and new was normally
2. create canisters from old wasm and create posts and bets. ideally multiple rooms
3. fetch the bet details of post (get_hot_or_not_bets_placed_by_this_profile_with_pagination and get_hot_or_not_bet_details_for_this_post)
4. upgrade to new wasm
5. fetch same bet details and match. 
6. add few more bets to see if all is working fine

Similar example tests are bet_details_heap_to_stable_mem_upgrade and feed_filter_upgrade_test 

This test has to be run and verified locally and will be #[ignore] in main.

------------------------

One other thing we have to check is whether these changes will impact hotornot site in any way.

File: src/lib/integration_tests/tests/hot_or_not/hotornot_game_simulation_test.rs [hunk 3]
```diff
@@ -526,8 +526,16 @@ fn hotornot_game_simulation_test() {
...
+    pic.advance_time(Duration::from_secs(60 * 60 + 30));
+    for _i in 1..=5 {
+        pic.tick();
+    }
+
+    // Forward timer for another post
+    pic.advance_time(Duration::from_secs(10));
+    for _i in 1..=5 {
+        pic.tick();
+    }
```
[the remark this answers is not in the document]
Comment (TwistingTwists) [reply src/lib/integration_tests/tests/hot_or_not/hotornot_game_simulation_test.rs:538]: Somehow, doing it twice gives different result than doing it once.

File: src/canister/platform_orchestrator/can.did [hunk 4]
```diff
@@ -6,11 +6,11 @@ type CanisterUpgradeStatus = record {
 type HttpRequest = record {
   url : text;
   method : text;
-  body : vec nat8;
+  body : blob;
```
[the remark this answers is not in the document]
Comment (ravibazz) [reply src/canister/platform_orchestrator/can.did:9]: Yes candid changed to blob. This is good

File: src/lib/shared_utils/src/canister_specific/individual_user_template/types/hot_or_not/mod.rs [hunk 5]
```diff
@@ -826,14 +1038,16 @@ impl Post {
             VirtualMemory<DefaultMemoryImpl>,
         >,
     ) {
+        let slot_id_owned = slot_id_type.clone();
+
         let hot_or_not_details = self.hot_or_not_details.as_mut();
 
         if hot_or_not_details.is_none() {
             return;
         }
```
Comment (komal-rs) [inline src/lib/shared_utils/src/canister_specific/individual_user_template/types/hot_or_not/mod.rs:1047]: is this relevant now or should we proceed to tabulate with inserting Some default ? In the context of old posts which were hotornotdetails - None

**Review by komal-sai-yral — COMMENTED** [review]
done reviewing
-----------
One thing to be checked is old posts which have hot_or_not_details as None, will that impact anything if we vote on it. On my side checked and looks ok except 1 tabulate_scores issue highlghted. Do check from your side once as well

File: src/canister/individual_user_template/can.did [hunk 6]
```diff
@@ -30,7 +30,7 @@ type BettingStatus = variant {
   BettingOpen : record {
     number_of_participants : nat8;
     ongoing_room : nat64;
-    ongoing_slot : nat8;
+    ongoing_slot : nat64;
```
[the remark this answers is not in the document]
Comment (komal-rs) [reply src/canister/individual_user_template/can.did:33]: Written this in morning. I think we already discussed this regarding backward compatibility of APIs

**Review by komal-sai-yral — COMMENTED** [review]
here as well - hot_or_not_details.is_none()

```
pub fn inform_participants_of_outcome(
    post: &Post,
    slot_id_type: &NewSlotType,
    room_details_map: &ic_stable_structures::btreemap::BTreeMap<
        GlobalRoomId,
        RoomDetailsV1,
        VirtualMemory<DefaultMemoryImpl>,
    >,
    bet_details_map: &ic_stable_structures::btreemap::BTreeMap<
        GlobalBetId,
        BetDetails,
        VirtualMemory<DefaultMemoryImpl>,
    >,
) {
    let slot_id_owned = slot_id_type.clone();

    let hot_or_not_details = post.hot_or_not_details.as_ref();

    if hot_or_not_details.is_none() {
        return;
    }

```

**Pull request closed without merging**